contracts/protocol/modules/v1/MorphoLeverageModule.sol

```diff
@@ -0,0 +1,920 @@
+    {
+        (uint256 collateralBalance, uint256 borrowBalance, ) = _getCollateralAndBorrowBalances(_setToken, _marketParams);
+        collateralPosition = collateralBalance.preciseDiv(_setTotalSupply).toInt256();
+        borrowPosition = borrowBalance.preciseDivCeil(_setTotalSupply).toInt256().mul(-1);
+    }
+
+    /**
+     * @dev Reads outstanding borrow token debt and collateral token balance from Morpho
+     */
+    function _getCollateralAndBorrowBalances(
+        ISetToken _setToken,
+        IMorpho.MarketParams memory _marketParams
+    )
+        internal
+        view 
+        returns(uint256 collateralBalance, uint256 borrowBalance, uint256 borrowSharesU256)
+    {
+        // TODO: Review if it is more efficient to generate this id once and save in storage
+        bytes32 marketId = _marketParams.id();
+        ( , , uint128 totalBorrowAssets, uint128 totalBorrowShares, , ) = morpho.market(marketId);
+        ( , uint128 borrowShares, uint128 collateral) = morpho.position(marketId, address(_setToken));
+        uint256 borrowAssets = borrowShares.toAssetsUp(totalBorrowAssets, totalBorrowShares);
+
+        collateralBalance = uint256(collateral);
+        borrowSharesU256 = uint256(borrowShares);
+        borrowBalance = borrowAssets;
+    }
+
+    /**
+     * @dev Updates external position unit for given borrow asset on SetToken
+     */
+    function _updateExternalPosition(
+        ISetToken _setToken, 
+        address _underlyingAsset,
+        int256 _newPositionUnit
+    )
+        internal
+    {
+        _setToken.editExternalPosition(_underlyingAsset, address(this), _newPositionUnit, "");
+    }
+
+
+    /**
+     * @dev Invoke deposit (as collateral) from SetToken using Morpho Blue
+     */
+    function _deposit(
+        ISetToken _setToken,
+        IMorpho.MarketParams memory _marketParams,
+        uint256 _notionalQuantity
+    )
+        internal
+    {
+        _setToken.invokeApprove(_marketParams.collateralToken, address(morpho), _notionalQuantity);
+        _setToken.invokeSupplyCollateral(
+            morpho,
+            _marketParams,
+            _notionalQuantity
+        );
+    }
+
+    /**
+     * @dev Invoke withdraw from SetToken using Morpho Blue
+     */
+    function _withdraw(
+        ISetToken _setToken,
+        IMorpho.MarketParams memory _marketParams,
+        uint256 _notionalQuantity
+    )
+        internal
+    {
+        _setToken.invokeWithdrawCollateral(
+            morpho,
+            _marketParams,
+            _notionalQuantity
+        );
+    }
+
+    /**
+     * @dev Invoke repay from SetToken using Morpho Blue
+     */
+    function _repayBorrow(
+        ISetToken _setToken,
+        IMorpho.MarketParams memory _marketParams,
+        uint256 _notionalQuantity,
+        uint256 _shares
+    )
+        internal
+    {
+        // Only ever set shares or assets to avoid "inconsistent input" error when there is a rounding error
+        if(_shares > 0) {
```

Comment:
Using shares is recommended when you want to fully close a position. It avoids leaving dust on Morpho.